Focused-Launch: Fix scroll to top when transitioning between routes

StefanNieuwenhuis opened the pull request:
We want each view in the Focused Launch to scroll to the top when transitioning between routes since information is "hidden" from the user's view when the scrolling position is preserved. This PR introduces a ScrollToTop component that is triggered on every route transition in the Focused Launch. It does the heavy-scroll-to-top-lifting on every route transition by hooking into the react-router-doms features.

#### Changes proposed in this Pull Request

* Every view in the Focused Launch should be scrolled to the top of the page automatically

#### Testing instructions
* Visit `/page/SANDBOXED_UNLAUNCHED_SITE_ID_CREATED_WITH_START.wordpress.com/home?flags=create/focused-launch-flow-calypso`. Focused Launch modal should appear
* Re-size the window reducing the height so a vertical scrollbar appears.
* Scroll down as far as you can, but make sure "View all domains" is still visible.
* Click on the _"View all domains"_ link. The new view is scrolled to the top of the page, showing all information.

Fixes #47325

StefanNieuwenhuis committed:
Focused-Launch: Fix scroll to top when transitioning between routes

We want each view in the Focused Launch to scroll to the top when transitioning between routes since information is "hidden" from the users view when the scrolling position is preserved. This PR introduces a ScrollToTop component that is triggered on every route transition in the Focused Launch. It does the heavy-scroll-to-top-lifting on every route transition by hooking in to the react-router-doms features.

StefanNieuwenhuis committed:
Focused-Launch: Fix failing CircleCI Build

tjcafferkey commented on packages/onboarding/src/scroll-to-top/index.tsx:
```diff
@@ -0,0 +1,17 @@
+const ScrollToTop: React.FunctionComponent = () => {
+	const { pathname } = useLocation();
+
+	React.useEffect( () => document.querySelector( '.components-modal__content' )?.scrollTo( 0, 0 ), [
```
Could we pass the selector as a prop in to the component to increase the flexibility of this component?

`<ScrollToTop selector=".components-modal__content" />`

tjcafferkey commented on packages/onboarding/src/scroll-to-top/test/index.tsx:
```diff
@@ -0,0 +1,30 @@
+/**
```
Good to see tests!

yansern replied:
+1

tjcafferkey commented on packages/onboarding/src/scroll-to-top/index.tsx:
```diff
@@ -0,0 +1,17 @@
+import * as React from 'react';
+import { useLocation } from 'react-router-dom';
+
+const ScrollToTop: React.FunctionComponent = () => {
```
Seeing as we only scroll to top for [specific viewports in Gutenboarding](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/onboarding-block/acquire-intent/vertical-select/index.tsx#L184) I wonder if it would be worth passing in viewports this would apply to (this would default to all viewports) and use the [useViewportMatch](https://github.com/Automattic/wp-calypso/blob/trunk/packages/domain-picker/src/domain-picker/suggestion-item.tsx#L64) hook

`<ScrollToTop viewport={['small', '>']} />`

And then we could use it like so:

`const matchesViewport = viewport ? useViewportMatch( viewport?.join(',') ) : true;`

Just a thought to make it more flexible although we could do this another time as we don't need this feature right now and is a little scope-creep.

StefanNieuwenhuis replied:
Hmm, I agree that it's a bit of a scope creep. I've created an issue ([#48161](https://github.com/Automattic/wp-calypso/issues/48161)) for this.

StefanNieuwenhuis committed:
Focused-Launch: Add selector prop to component

We want the component to be more flexible, hence we have added a selector property.

tjcafferkey reviewed (approved):
LGTM 👍🏻

StefanNieuwenhuis committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into fix/focused-launch/scroll-to-top

razvanpapadopol commented on packages/onboarding/src/scroll-to-top/index.tsx:
```diff
@@ -0,0 +1,22 @@
+		pathname,
+	] );
+
+	return null;
```
Why aren't we using a custom hook instead of having a component returning `null`?

tjcafferkey replied:
Good point/spot @razvanpapadopol. @StefanNieuwenhuis could we update this to become a hook?

StefanNieuwenhuis replied:
We're unable to use a custom hook in the `<Router/>` and in order to comply with best practices, we have to use a component (see [react-router-dom docs](https://reactrouter.com/web/guides/scroll-restoration/scroll-to-top)). 

In other use-cases a hook absolutely makes sense but that's out of scope for this issue.

razvanpapadopol commented on packages/onboarding/src/scroll-to-top/index.tsx:
```diff
@@ -0,0 +1,22 @@
+const ScrollToTop: React.FunctionComponent< ScrollToTopProps > = ( { selector } ) => {
+	const { pathname } = useLocation();
+
+	React.useEffect( () => document.querySelector( selector )?.scrollTo( 0, 0 ), [
```
While I understand the flexibility of having a prop such as `selector` here, I have two suggestions:
- can we make it optional and when left undefined just use `window.scrollTo()`?
- for greater stability can we use a `ref` instead? ([example](https://levelup.gitconnected.com/react-tips-scroll-to-element-load-fonts-run-setstate-in-sequence-2d73c9f02ae1#d959))

StefanNieuwenhuis replied:
> can we make it optional and when left undefined just use `window.scrollTo()`?

Done ✅

> for greater stability can we use a `ref` instead? 

We can use a `ref` but that means that we have to pass it all the way from the parent's parent. We have to convert them to `React.forwardRef()` and I'm unsure what the consequences are. Besides that, we have to wrap the `<Router/>` in `<ScrollToTop/>` which is against best practices (see [react-router docs](https://reactrouter.com/web/guides/scroll-restoration/scroll-to-top)). 

To conclude: It becomes ugly pretty fast and we cannot oversee the consequences. 
We're going for the selector option w/ a fallback to `window.scrollTo()`.

StefanNieuwenhuis committed:
Focused-Launch: Add window.scrollTo fallback

razvanpapadopol reviewed (approved):
🚀
